modules/imgcodecs/src/grfmt_png.cpp

```diff
@@ -449,6 +941,824 @@ bool  PngEncoder::write( const Mat& img, const std::vector<int>& params )
+        sub_row = new uchar[rowbytes + 1];
+        up_row = new uchar[rowbytes + 1];
+        avg_row = new uchar[rowbytes + 1];
+        paeth_row = new uchar[rowbytes + 1];
+
+        row_buf[0] = 0;
+        sub_row[0] = 1;
+        up_row[0] = 2;
+        avg_row[0] = 3;
+        paeth_row[0] = 4;
+
+        x0 = 0;
+        y0 = 0;
+        w0 = width;
+        h0 = height;
+        bop = 0;
+        next_seq_num = 0;
+
+        for (j = 0; j < 6; j++)
+            op[j].valid = 0;
+        deflate_rect_op(frames[0].pixels(), x0, y0, w0, h0, bpp, rowbytes, zbuf_size, 0);
+        deflate_rect_fin(deflate_method, iter, zbuf, &zsize, bpp, rowbytes, rows, zbuf_size, 0);
+
+        if (first)
+        {
+            write_IDATs(f, 0, zbuf, zsize, idat_size);
+            for (j = 0; j < 6; j++)
+                op[j].valid = 0;
+            deflate_rect_op(frames[1].pixels(), x0, y0, w0, h0, bpp, rowbytes, zbuf_size, 0);
+            deflate_rect_fin(deflate_method, iter, zbuf, &zsize, bpp, rowbytes, rows, zbuf_size, 0);
+        }
+
+        for (i = first; i < num_frames - 1; i++)
+        {
+            uint op_min;
+            int op_best;
+
+            for (j = 0; j < 6; j++)
+                op[j].valid = 0;
+
+            /* dispose = none */
+            get_rect(width, height, frames[i].pixels(), frames[i + 1].pixels(), over1, bpp, rowbytes, zbuf_size, has_tcolor, tcolor, 0);
+
+            /* dispose = background */
+            if (has_tcolor)
+            {
+                memcpy(temp, frames[i].pixels(), imagesize);
+                if (coltype == 2)
+                    for (j = 0; j < h0; j++)
+                        for (k = 0; k < w0; k++)
+                            memcpy(temp + ((j + y0) * width + (k + x0)) * 3, &tcolor, 3);
+                else
+                    for (j = 0; j < h0; j++)
+                        memset(temp + ((j + y0) * width + x0) * bpp, tcolor, w0 * bpp);
+
+                get_rect(width, height, temp, frames[i + 1].pixels(), over2, bpp, rowbytes, zbuf_size, has_tcolor, tcolor, 1);
+            }
+
+            /* dispose = previous */
+            if (i > first)
+                get_rect(width, height, rest, frames[i + 1].pixels(), over3, bpp, rowbytes, zbuf_size, has_tcolor, tcolor, 2);
+
+            op_min = op[0].size;
+            op_best = 0;
+            for (j = 1; j < 6; j++)
+                if (op[j].valid)
+                {
+                    if (op[j].size < op_min)
+                    {
+                        op_min = op[j].size;
+                        op_best = j;
+                    }
+                }
+
+            dop = op_best >> 1;
+
+            png_save_uint_32(buf_fcTL, next_seq_num++);
+            png_save_uint_32(buf_fcTL + 4, w0);
+            png_save_uint_32(buf_fcTL + 8, h0);
+            png_save_uint_32(buf_fcTL + 12, x0);
+            png_save_uint_32(buf_fcTL + 16, y0);
+            png_save_uint_16(buf_fcTL + 20, frames[i].delayNum());
```

Comment:
Some of those members are unset right now because the user does not send them. The writemulti API is not good enough to write an animation.